src/query/query_parse.rs

```diff
@@ -230,22 +301,103 @@ mod tests {
         let res = parse_query(r#"id >= "123""#).unwrap();
-        assert!(matches!(res, QueryNode::Test(Test::GreaterThanOrEqual, Lookup::Id, Literal::String("123"))));
+        assert_node!(res, QueryNode::Test(Test::GreaterThanOrEqual, Lookup::Id, Literal::String("123")));
 
         let e = parse_query(r#"name =! "123""#).unwrap_err();
-        assert!(matches!(e, QueryError::ParseError(_)));
+        assert_node!(e, QueryError::ParseError(_));
         assert!(e
             .to_string()
             .contains("Encountered unknown infix operator: =!"));
     }
+
+    #[test]
+    fn query_can_parse_negation_and_brackets() {
+        let res = parse_query("!latest()").unwrap();
+        assert_node!(res, QueryNode::Negation, QueryNode::Latest(None));
+
+        let res = parse_query("(latest())").unwrap();
+        assert_node!(res, QueryNode::Brackets, QueryNode::Latest(None));
+
+        let res = parse_query(r#"id == "123""#).unwrap();
+        assert_node!(res, QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("123")));
+
+        let res = parse_query(r#"!id == "123""#).unwrap();
+        assert_node!(res, QueryNode::Negation,
+            QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("123")));
+
+        let res = parse_query(r#"(!id == "123")"#).unwrap();
+        assert_node!(res, QueryNode::Brackets, QueryNode::Negation,
+            QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("123"))
+        );
+
+        let res = parse_query(r#"!(!id == "123")"#).unwrap();
+        assert_node!(res, QueryNode::Negation, QueryNode::Brackets, QueryNode::Negation,
+            QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("123"))
+        );
+    }
+
+    #[test]
+    fn query_can_parse_logical_operators() {
+        let res = parse_query(r#"id == "123" || id == "345""#).unwrap();
+        assert_node!(res, QueryNode::BooleanOperator, Operator::Or,
+            QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("123")),
+            QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("345")));
+
+        let res = parse_query(r#"id == "123" && id == "345""#).unwrap();
+        assert_node!(res, QueryNode::BooleanOperator, Operator::And,
+            QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("123")),
+            QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("345")));
+
+        let res = parse_query(r#"id == "123" && id == "345" || id == "this""#).unwrap();
+        match res {
+            QueryNode::BooleanOperator(Operator::Or, value1, value2) => {
+                assert_node!(*value1, QueryNode::BooleanOperator, Operator::And,
+                    QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("123")),
+                    QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("345")));
+                assert!(matches!(*value2, QueryNode::Test(Test::Equal, Lookup::Id, Literal::String("this"))));
```

Comment:
feels like this can be put into a macro but i couldnt quite get it working, im assuming this is what you were stuck on? its a bit annoying, will maybe come back to this when there is more time XD guess this macro is not full recursion!